diff --git a/bin/generate/generate.js b/bin/generate/generate.js
index 944fffe..40448e1 100644
--- a/bin/generate/generate.js
+++ b/bin/generate/generate.js
@@ -4,6 +4,7 @@ import ssr from '../../ssr/index.js'
 import { wrap } from '../../ssr/shared.js'
 
 import config, { resolve } from '../config.js'
+import { resolveLocation } from './shared.js'
 
 await fs.rm(config.outputDir, { recursive: true, force: true })
 await fs.mkdir(config.outputDir, { recursive: true })
@@ -19,7 +20,10 @@ await generate()
 console.log('Finished generating', files, 'files in', (performance.now() - start).toFixed(1) + 'ms') // eslint-disable-line
 process.exit(process.exitCode)
 
-async function generate(location = '/') {
+async function generate(href = '/', from) {
+  const url = resolveLocation(href, from)
+      , location = url.pathname + url.search
+
   if (generated.has(location))
     return
 
@@ -34,5 +38,5 @@ async function generate(location = '/') {
   await fs.mkdir(path.dirname(indexPath), { recursive: true, force: true })
   await fs.writeFile(indexPath, html)
   files++
-  await Promise.all([...(x.links || [])].map(generate))
+  await Promise.all([...(x.links || [])].map(href => generate(href, url)))
 }
diff --git a/bin/generate/shared.js b/bin/generate/shared.js
new file mode 100644
index 0000000..4992c03
--- /dev/null
+++ b/bin/generate/shared.js
@@ -0,0 +1,5 @@
+export function resolveLocation(href, from = 'http://localhost/') {
+  const location = new URL(href, from)
+  location.hash = ''
+  return location
+}
diff --git a/src/index.js b/src/index.js
index ec2733d..ed81827 100644
--- a/src/index.js
+++ b/src/index.js
@@ -419,16 +419,16 @@ function shouldHydrate(dom, context, attrs) {
   if (dom) {
     if (dom.nodeType === 8 && dom.data === 'h') {
       dom.remove()
-      context.hydrate = true
+      context.hydrating = true
     } else if (dom.nodeType === 1 && dom.tagName === 'SCRIPT' && dom.hasAttribute('h')) {
-      context.hydrate = true
+      context.hydrating = true
       const x = JSON.parse(dom.textContent)
       Object.assign(context, x.context)
       Object.assign(attrs, x.attrs)
     }
   }
 
-  if (!context.hydrate)
+  if (!context.hydrating)
     return
 
   let node
diff --git a/ssr/index.js b/ssr/index.js
index 8537271..d608f65 100644
--- a/ssr/index.js
+++ b/ssr/index.js
@@ -190,6 +190,7 @@ function openingTag(view, tag) {
     + tag
     + getClassName(view)
     + Object.entries(view.attrs).reduce((acc, [k, v]) => acc += renderAttr(k, v), '')
+    + Object.entries(view.attrs.data || emptyObject).reduce((acc, [k, v]) => acc += renderAttr('data-' + styleProp(k), v), '')
     + getStyle(view)
     + '>'
 }
@@ -264,6 +265,7 @@ function updateComment(view) {
 }
 
 function updateComponent(view, context) {
+  context = view.component[1] && view.component[1].context || context
   let x = view.component[0](view.attrs, view.children, context)
   mergeTag(x, view)
   return tryPromise(x, (x, wasPromise) => {
